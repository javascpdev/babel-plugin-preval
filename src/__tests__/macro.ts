import path from 'path'
import pluginTester from 'babel-plugin-tester'
import plugin from 'babel-plugin-macros'

const projectRoot = path.join(__dirname, '../../')

expect.addSnapshotSerializer({
  print(val) {
    const valString = val as string
    return valString.split(projectRoot).join('<PROJECT_ROOT>/')
  },
  test(val) {
    return typeof val === 'string'
  },
})

pluginTester({
  plugin,
  pluginName: 'preval',
  snapshot: true,
  babelOptions: {filename: __filename, parserOpts: {plugins: ['jsx']}},
  tests: [
    {
      title: 'as tag',
      code: `
        import preval from './helpers/preval.macro'

        const x = preval\`module.exports = require('./fixtures/compute-one')\`
      `,
    },
    {
      title: 'as function',
      code: `
        const myPreval = require('./helpers/preval.macro')

        const x = myPreval(\`
          module.exports = require('./fixtures/identity')({sayHi: () => 'hi'})
        \`)
      `,
    },
    {
      title: 'as jsx',
      code: `
        const Preval = require('./helpers/preval.macro')

        const ui = (
          <Preval>
            const fs = require('fs')
            module.exports = fs.readFileSync(require.resolve('./fixtures/fixture1.md'), 'utf8')
          </Preval>
        )
      `,
    },
    {
      title: 'error for other nodes',
      error: true,
      code: `
        const preval = require('./helpers/preval.macro')

        x = 3 + preval
      `,
    },
  ],
})
