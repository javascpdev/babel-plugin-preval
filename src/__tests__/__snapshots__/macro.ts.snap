// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preval as function: as function 1`] = `

const myPreval = require('./helpers/preval.macro')

const x = myPreval(\`
  module.exports = require('./fixtures/identity')({sayHi: () => 'hi'})
\`)

      ↓ ↓ ↓ ↓ ↓ ↓

const x = {
  sayHi: () => 'hi',
}


`;

exports[`preval as jsx: as jsx 1`] = `

const Preval = require('./helpers/preval.macro')

const ui = (
  <Preval>
    const fs = require('fs')
    module.exports = fs.readFileSync(require.resolve('./fixtures/fixture1.md'), 'utf8')
  </Preval>
)

      ↓ ↓ ↓ ↓ ↓ ↓

const ui = <div>"# fixture\\n\\nThis is some file thing...\\n"</div>


`;

exports[`preval as tag: as tag 1`] = `

import preval from './helpers/preval.macro'

const x = preval\`module.exports = require('./fixtures/compute-one')\`

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 1


`;

exports[`preval error for other nodes: error for other nodes 1`] = `

const preval = require('./helpers/preval.macro')

x = 3 + preval

      ↓ ↓ ↓ ↓ ↓ ↓

Error: <PROJECT_ROOT>/src/__tests__/macro.ts: ./helpers/preval.macro: babel-plugin-preval/macro can only be used as tagged template expression, function call or JSX element. You tried BinaryExpression.

`;
